frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1510,11 +1524,47 @@ const IR::Node *TypeInference::postorder(IR::SerEnumMember *member) {
     */
     auto serEnum = findContext<IR::Type_SerEnum>();
     CHECK_NULL(serEnum);
-    auto type = getTypeType(serEnum->type);
-    if (!type->is<IR::Type_Bits>()) {
-        typeError("%1%: Illegal type for enum; only bit<> and int<> are allowed", serEnum->type);
+    const auto *resolvedType = getTypeType(serEnum->type);
+    const auto *type = resolvedType->to<IR::Type_Bits>();
+    if (!type) {
+        std::string note;
+        if (resolvedType->is<IR::Type_InfInt>()) {
+            note = "; note that the used type is unsized integral type";
+        }
+        if (resolvedType->is<IR::Type_Newtype>()) {
+            note = "; note that type-declared types are not allowed even if they are fixed-size";
+        }
+        typeError("%1%: Illegal type for enum; only bit<> and int<> are allowed%2%", serEnum->type,
+                  note);
         return member;
     }
+    // validate the constant fits -- non-fitting enum constants should pruce error
+    if (const auto *constant = member->value->to<IR::Constant>()) {
```

Comment:
I would have implemented this by casting the value to the underlying type, that should give a warning already.
But I guess this works.

Reply:
I think emitting the warning and then somehow processing it and turning it into error would be more confusing.